### ISSUE TITLE
explicit `x is None` should narrow union candidates like `isinstance(x, type(None))`

### ISSUE THREAD

**Issue by bwo** [python/mypy#4520]
```
from typing import Optional, List  # noqa


def union_test1(x):
    # type: (Optional[List[int]]) -> Optional[int]
    if x is None:
        return x
    else:
        return len(x)


def union_test2(x):
    # type: (Optional[List[int]]) -> Optional[int]
    if isinstance(x, type(None)):
        return x
    else:
        return len(x)
```

`union_test1` with default options gives the error `Incompatible return value type (got "Optional[List[int]]", expected "Optional[int]")` (with `--strict-optional`, it does not give this error). `union_test2` typechecks just fine. This strikes me as highly counterintuitive; even though (as the docs say) `None` is a valid value for each type, if one of the above checks allows the typechecker to learn that `x` is `None` and *therefore* not a list, it's hard to see why the other doesn't.

Failing that, I would suggest at least a documentation change to point out both the gotcha embodied in `union_test1` and the workaround in `union_test2`.

**Comment by gvanrossum** [comment]
Agreed, I don't see why the first one isn't treated the same as the second.

**Comment by ilevkivskyi** [comment]
> Agreed, I don't see why the first one isn't treated the same as the second.

This is not the first time I see weird behaviour of `None` _without_ `--strict-optional`.

**Comment by msullivan** [comment]
It looks like the code in `find_isinstance_check` to refine optionals based on comparisons is explicitly gated on `--strict-optional`: https://github.com/python/mypy/blob/master/mypy/checker.py#L2865

Is there any reason to not just change that?
